### PR TITLE
WIP SimpleNamespace

### LINKED ISSUE #1645: Support SimpleNamespace
Currently `SimpleNamespace` doesn't populate autocomplete suggestions for set attributes. I've looked through the issues and haven't seen `SimpleNamespace` mentioned. Could be a similar problem to `Enum`, `namedtuple`, or `dataclass`?

FWIW I see the [same issue](https://github.com/microsoft/python-language-server/issues/2108) when using the Microsoft Python Language Server in VSCode, but Jedi actually gives me different Intellisense options if you compare the screenshots between the two issues.

## Environment data
VSCode

> Version: 1.47.2
> Commit: 17299e413d5590b14ab0340ea477cdd86ff13daf
> Date: 2020-07-15T18:18:50.054Z
> Electron: 7.3.2
> Chrome: 78.0.3904.130
> Node.js: 12.8.1
> V8: 7.8.279.23-electron.0
> OS: Darwin x64 18.7.0

Python Extension

> Name: Python
> Id: ms-python.python
> Version: 2020.7.96456
> Publisher: Microsoft

Language Server Settings
```
{
  "python.languageServer": "Jedi",
}
```

## Code Example
```python
from types import SimpleNamespace
colors = SimpleNamespace(red='red', blue='blue')
# this should autocomplete, it doesn't find `red` or `blue` as options
colors.red
```

### Screenshot
image

### PR DESCRIPTION
**Pull request by kbehlers**
Work In Progress
SimpleNamespace completion

Closes #1645

### CODE REVIEW

**Comment by davidhalter** [comment]
You are on the right path. The development docs probably won't help you too much with this issue. Sorry for that, but it hasn't really yielded a lot of value in the last few years, so I have been neglecting it.

You should be using something like (I'm paraphrasing, not sure this is completely correct):

```
def get_filters(self, origin_scope=None):
    def iterate():
        for key, lazy_value in self._arguments.unpack():
            if key is not None:
                yield key, SimpleNamespaceName(self.inference_state, key, lazy_value)
    yield DictFilter(dict(iterate()))

class SimpleNamespaceName(AbstractArbitraryName):
    def __init__(self, inference_state, string, lazy_value):
        super().__init__(inference_state, string)
        self._lazy_value = lazy_value

    def infer(self):
        return self._lazy_value.infer()
```

This method is obviously in the SimpleNamespaceWrapper.

Not sure this all works. Obviously also needs tests. But it's something like that I guess. In general you could also overwrite other methods than `get_filters`, but get_filters allows you to get completions and type inference working. Goto does not work this way, but I guess that's fine. To support that we would probably need to rewrite quite a bit of `unpack`.

This might feel like a lot of help for this, but most things are trivial if you know how the abstractions work, I just did not want to leave you guessing all of this. :)

**Comment by kbehlers** [comment]
@davidhalter Thank you! I'll try to implement what you suggested. I read through the developer documentation, and I know I am not understanding all the abstractions for Jedi.

Initially I had hoped I could closely model `dataclass`, but again I don't understand enough just yet.
https://github.com/davidhalter/jedi/blob/8565adad91d12e34d4724aa4321763984d522135/jedi/plugins/stdlib.py#L601-L634

This is the research I had done, and I'll still post it in case any of it proves helpful, but it sounds like your suggestion may get me where I am trying to go.


Here's what I know about `SimpleNamespace`
- It is implemented in C [here](https://github.com/python/cpython/blob/6b6092f533f0e4787b8564c4fad6ec6d1018af0d/Objects/namespaceobject.c)
- Every attribute we would want to code complete is stored in the underlying `__dict__` so:
```
>>>from types import SimpleNamespace
>>>s = SimpleNamespace(example='hello')
>>>vars(s)
{'example': 'hello}
```
- `SimpleNamespace` is made available in `types` [here](https://github.com/python/cpython/blob/master/Lib/types.py#L16) as so:
```
SimpleNamespace = type(sys.implementation)
```
- Jedi seems to follow that for completion to [here](https://github.com/davidhalter/typeshed/blob/d38645247816f862cafeed21a8f4466d306aacf3/stdlib/3/sys.pyi#L110-L115)
```
# Stubs for sys
implementation: _implementation
class _implementation:
    name: str
    version: _version_info
    hexversion: int
    cache_tag: str
```
- `sys.implementation` is also written in C and basically wraps SimpleNamespace [here](https://github.com/python/cpython/blob/master/Python/sysmodule.c#L2630), which is why 
```
>>> type(sys.implementation)
<class 'types.SimpleNamespace'>
```
So most of those details probably don't matter, but maybe some will be relevant.

File: jedi/plugins/stdlib.py [hunk 1]
```diff
@@ -591,11 +591,31 @@ def _random_choice(sequences):
...
+    # TODO copied roughly from dataclass, obviously wrong
+    for c in _follow_param(value.inference_state, arguments, 0):
+        if c.is_class():
+            return ValueSet([SimpleNamespaceWrapper(c)])
+        else:
+            return ValueSet([value])
+    return NO_VALUES
 
 
 class SimpleNamespaceWrapper(ValueWrapper, ClassMixin):
```
[the remark this answers is not in the document]
Comment (kbehlers) [reply jedi/plugins/stdlib.py]: Thank you!

**Comment by kbehlers** [comment]
@davidhalter I am wanting to be more helpful than I am probably capable of being given my limited understanding of Jedi.

Could you help me understand how `_simple_namespace` can be wired up to `SimpleNamespaceWrapper`? https://github.com/davidhalter/jedi/pull/1649/files#diff-193de5c70694b9299f09bd45ac2f70d1R594

File: jedi/plugins/stdlib.py [hunk 2]
```diff
@@ -594,7 +594,8 @@ def _simple_namespace(value, arguments, callback):
     # TODO how do I incorporate SimpleNamespaceWrapper correctly?
     return ValueSet([SimpleNamespaceWrapper(value, arguments)])
 
-class SimpleNamespaceWrapper(ValueWrapper, ClassMixin):
+
+class SimpleNamespaceWrapper(Value):
     def __init__(self, cls, arguments):
```
Comment (davidhalter) [inline jedi/plugins/stdlib.py:599]: just passing inference_state and arguments should be good enough.

**Comment by davidhalter** [comment]
Not sure if you are waiting for input from my side, but this definitely needs tests.

Also don't worry about Python 3.9 tests failing.

**Comment by kbehlers** [comment]
@davidhalter Let me try to get some tests and that might get me going on this again.

**Comment by kbehlers** [comment]
@davidhalter Revisiting this, I'm just not grokking the codebase like I had hoped and the ["Is it worth the time?"](https://xkcd.com/1205/) numbers aren't going to work out for me.

How do you want me to "cleanup" at this point? Close the PR? Close the Issue as well?

**Comment by davidhalter** [comment]
Haha :) Yeah, if you don't want to invest the time, just close both issue and PR. I don't think it would have taken way longer, but I can understand that it's not worth it for you personally.

**Pull request closed without merging**